File: app/lib/e2e/channel.dart
import 'dart:convert';
import 'dart:typed_data';

import '../transport/jsonrpc.dart' show RpcError, RpcMessage;
import 'session.dart';

/// Method for a relay frame carrying a raw Noise handshake message (msg1/msg2),
/// before any sealed app traffic.
const String methodE2EHandshake = 'e2e.handshake';

/// The Noise prologue binding a channel to its node id and chan id. Client and
/// node MUST derive it identically.
Uint8List channelPrologue(String nodeId, String chanId) =>
    Uint8List.fromList(utf8.encode('argus-e2e/v1|$nodeId|$chanId'));

/// Cleartext routing metadata a blind gateway reads to relay a frame.
class RouteHeader {
  const RouteHeader({required this.chanId, this.nodeId, this.subId, this.termId});

  final String chanId;
  final String? nodeId;
  final String? subId;
  final String? termId;

  static RouteHeader? fromJson(Object? j) {
    if (j is! Map) return null;
    final chan = j['chan_id'];
    if (chan is! String) return null;
    return RouteHeader(
      chanId: chan,
      nodeId: j['node_id'] as String?,
      subId: j['sub_id'] as String?,
      termId: j['term_id'] as String?,
    );
  }
}

/// A parsed relay frame. [body] is the JSON body value (a base64 String for
/// sealed or handshake frames); [raw] is the verbatim line.
class RelayFrame {
  RelayFrame({this.method, this.id, this.route, this.body, required this.raw});

  final String? method;
  final Object? id;
  final RouteHeader? route;
  final Object? body;
  final Uint8List raw;

  static RelayFrame parse(List<int> line) {
    final data = line is Uint8List ? line : Uint8List.fromList(line);
    final j = jsonDecode(utf8.decode(data)) as Map<String, dynamic>;
    return RelayFrame(
      method: j['method'] as String?,
      id: j['id'],
      route: RouteHeader.fromJson(j['route']),
      body: j['body'],
      raw: data,
    );
  }

  static RelayFrame fromMessage(RpcMessage m) => RelayFrame(
        method: m.method,
        id: m.id,
        route: RouteHeader.fromJson(m.route),
        body: m.body,
        raw: Uint8List(0),
      );
}

/// Builds a relay frame carrying a raw (unsealed) Noise handshake message.
Uint8List marshalHandshakeFrame(String chanId, List<int> handshake) {
  final frame = <String, Object?>{
    'jsonrpc': '2.0',
    'method': methodE2EHandshake,
    'route': {'chan_id': chanId},
    'body': base64.encode(handshake),
  };
  return Uint8List.fromList(utf8.encode(jsonEncode(frame)));
}

/// Extracts the raw Noise handshake bytes from a handshake frame.
Uint8List handshakeFromFrame(RelayFrame f) {
  final b = f.body;
  if (b is! String) {
    throw const FormatException('handshake frame body is not a string');
  }
  return Uint8List.fromList(base64.decode(b));
}

/// Seals JSON-RPC payloads for one client<->node E2E channel. The routing header
/// travels in cleartext so a blind gateway can relay by chan_id; the inner
/// params/result/error travel sealed in the body.
class Channel {
  Channel(this.chanId, this._session);

  final String chanId;
  final Session _session;

  String _sealBody(List<int> inner) => base64.encode(_session.seal(inner));

  Uint8List _openBody(Object? body) {
    if (body is! String) {
      throw const FormatException('channel body is not a string');
    }
    return _session.open(base64.decode(body));
  }

  /// Seals a request into wire frame bytes. [params] is the raw params JSON.
  Uint8List sealRequestFrame(int id, String method, String nodeId, List<int> params) {
    final route = <String, Object?>{'chan_id': chanId};
    if (nodeId.isNotEmpty) route['node_id'] = nodeId;
    final frame = <String, Object?>{
      'jsonrpc': '2.0',
      'id': id,
      'method': method,
      'route': route,
      'body': _sealBody(params),
    };
    return Uint8List.fromList(utf8.encode(jsonEncode(frame)));
  }

  /// Decrypts a response frame's body into its raw result JSON bytes or rpc error.
  /// Relies on Go's canonical sealedResponse marshaling: `{"result":<raw>}` or
  /// `{"error":<raw>}` (no spaces, exactly one key). Returns raw bytes so int64
  /// results survive on the web (no JSON number decode).
  ({Uint8List? result, RpcError? error}) openResponse(RelayFrame f) {
    final inner = _openBody(f.body);
    if (_hasPrefix(inner, '{"error":')) {
      final m = jsonDecode(utf8.decode(inner)) as Map<String, dynamic>;
      return (result: null, error: RpcError.fromJson(m['error'] as Map<String, dynamic>));
    }
    const p = '{"result":';
    if (_hasPrefix(inner, p) && inner.last == 0x7d /* } */) {
      return (result: Uint8List.sublistView(inner, p.length, inner.length - 1), error: null);
    }
    throw const FormatException('unexpected sealed response shape');
  }

  static bool _hasPrefix(Uint8List b, String p) {
    if (b.length < p.length) return false;
    for (var i = 0; i < p.length; i++) {
      if (b[i] != p.codeUnitAt(i)) return false;
    }
    return true;
  }

  /// Decrypts a request/notification frame's body into its raw params JSON bytes.
  Uint8List openParams(RelayFrame f) => _openBody(f.body);
}

File: internal/api/channel.go
package api

import (
	"encoding/base64"
	"encoding/json"
	"fmt"

	"github.com/MunifTanjim/argus/internal/e2e"
)

// RelayFrame is a JSON-RPC frame on a relayed (gateway) link. A blind gateway
// routes it by Route.ChanID alone and forwards Raw verbatim to the paired peer,
// never reading the cleartext Method/ID or touching the sealed Body.
type RelayFrame struct {
	Method string
	ID     *json.RawMessage
	Route  RouteHeader
	Body   json.RawMessage
	Raw    []byte
}

// Cipher seals and opens channel payloads. Noise for E2EE; identity for plaintext.
type Cipher interface {
	Seal(plaintext []byte) ([]byte, error)
	Open(sealed []byte) ([]byte, error)
}

type noiseCipher struct{ s *e2e.Session }

func (c noiseCipher) Seal(p []byte) ([]byte, error) { return c.s.Seal(p) }
func (c noiseCipher) Open(b []byte) ([]byte, error) { return c.s.Open(b) }

type identityCipher struct{}

func (identityCipher) Seal(p []byte) ([]byte, error) { return p, nil }
func (identityCipher) Open(b []byte) ([]byte, error) { return b, nil }

// Channel seals JSON-RPC payloads for one client<->node channel. The routing
// header travels in cleartext so a blind gateway can relay by chan_id; the inner
// params/result/error travel in the sealed, opaque Body.
type Channel struct {
	id     string
	cipher Cipher
}

func NewChannel(id string, session *e2e.Session) *Channel {
	return &Channel{id: id, cipher: noiseCipher{s: session}}
}

func NewPlainChannel(id string) *Channel {
	return &Channel{id: id, cipher: identityCipher{}}
}

func (c *Channel) ID() string { return c.id }

func (c *Channel) seal(inner json.RawMessage) (json.RawMessage, error) {
	sealed, err := c.cipher.Seal(inner)
	if err != nil {
		return nil, fmt.Errorf("api: channel seal: %w", err)
	}
	q, err := json.Marshal(base64.StdEncoding.EncodeToString(sealed))
	if err != nil {
		return nil, err
	}
	return json.RawMessage(q), nil
}

func (c *Channel) open(body json.RawMessage) (json.RawMessage, error) {
	var enc string
	if err := json.Unmarshal(body, &enc); err != nil {
		return nil, fmt.Errorf("api: channel body not a string: %w", err)
	}
	sealed, err := base64.StdEncoding.DecodeString(enc)
	if err != nil {
		return nil, fmt.Errorf("api: channel body base64: %w", err)
	}
	inner, err := c.cipher.Open(sealed)
	if err != nil {
		return nil, fmt.Errorf("api: channel open: %w", err)
	}
	return json.RawMessage(inner), nil
}

func (c *Channel) sealRequest(id *json.RawMessage, method, nodeID string, params json.RawMessage) (message, error) {
	body, err := c.seal(params)
	if err != nil {
		return message{}, err
	}
	return message{
		JSONRPC: jsonrpcVersion,
		ID:      id,
		Method:  method,
		Route:   &RouteHeader{ChanID: c.id, NodeID: nodeID},
		Body:    body,
	}, nil
}

// sealNotification always overwrites route.ChanID with this channel; the caller
// sets only SubID/TermID.
func (c *Channel) sealNotification(method string, route RouteHeader, params json.RawMessage) (message, error) {
	body, err := c.seal(params)
	if err != nil {
		return message{}, err
	}
	route.ChanID = c.id
	return message{
		JSONRPC: jsonrpcVersion,
		Method:  method,
		Route:   &route,
		Body:    body,
	}, nil
}

// sealedResponse is the inner (sealed) shape of a response: exactly one of Result
// or Error. Kept inside Body so node error detail never appears in cleartext.
type sealedResponse struct {
	Result json.RawMessage `json:"result,omitempty"`
	Error  *RPCError       `json:"error,omitempty"`
}

func (c *Channel) sealResponse(id *json.RawMessage, result json.RawMessage, rpcErr *RPCError) (message, error) {
	inner, err := json.Marshal(sealedResponse{Result: result, Error: rpcErr})
	if err != nil {
		return message{}, err
	}
	body, err := c.seal(inner)
	if err != nil {
		return message{}, err
	}
	return message{
		JSONRPC: jsonrpcVersion,
		ID:      id,
		Route:   &RouteHeader{ChanID: c.id},
		Body:    body,
	}, nil
}

func (c *Channel) OpenParams(f RelayFrame) (json.RawMessage, error) {
	return c.open(f.Body)
}

func (c *Channel) OpenResponse(f RelayFrame) (json.RawMessage, *RPCError, error) {
	inner, err := c.open(f.Body)
	if err != nil {
		return nil, nil, err
	}
	var r sealedResponse
	if err := json.Unmarshal(inner, &r); err != nil {
		return nil, nil, fmt.Errorf("api: response body: %w", err)
	}
	return r.Result, r.Error, nil
}

// SealRequestFrame output is ready for Peer.SendRawFrame.
func (c *Channel) SealRequestFrame(id *json.RawMessage, method, nodeID string, params json.RawMessage) ([]byte, error) {
	m, err := c.sealRequest(id, method, nodeID, params)
	if err != nil {
		return nil, err
	}
	return json.Marshal(m)
}

func (c *Channel) SealResponseFrame(id *json.RawMessage, result json.RawMessage, rpcErr *RPCError) ([]byte, error) {
	m, err := c.sealResponse(id, result, rpcErr)
	if err != nil {
		return nil, err
	}
	return json.Marshal(m)
}

// SealNotificationFrame overwrites route.ChanID with this channel; the caller sets
// only SubID/TermID.
func (c *Channel) SealNotificationFrame(method string, route RouteHeader, params json.RawMessage) ([]byte, error) {
	m, err := c.sealNotification(method, route, params)
	if err != nil {
		return nil, err
	}
	return json.Marshal(m)
}

// ParseRelayFrame mirrors the RelayFrame construction in Peer's read loop; keep
// them in sync.
func ParseRelayFrame(line []byte) (RelayFrame, error) {
	var m message
	if err := json.Unmarshal(line, &m); err != nil {
		return RelayFrame{}, fmt.Errorf("api: parse relay frame: %w", err)
	}
	if !m.isRelay() {
		return RelayFrame{}, fmt.Errorf("api: not a relay frame")
	}
	return RelayFrame{
		Method: m.Method,
		ID:     m.ID,
		Route:  *m.Route,
		Body:   m.Body,
		Raw:    append([]byte(nil), line...),
	}, nil
}

// MethodE2EHandshake carries a Noise handshake message (msg1/msg2) as a relay
// frame Body over a channel, before any sealed app traffic.
const MethodE2EHandshake = "e2e.handshake"

// ChannelPrologue is the Noise prologue binding a channel to its node id and chan
// id. Client and node MUST derive it identically.
func ChannelPrologue(nodeID, chanID string) []byte {
	return []byte("argus-e2e/v1|" + nodeID + "|" + chanID)
}

// MarshalHandshakeFrame carries a raw Noise handshake message unsealed — there is
// no session yet.
func MarshalHandshakeFrame(chanID string, handshake []byte) ([]byte, error) {
	body, err := json.Marshal(base64.StdEncoding.EncodeToString(handshake))
	if err != nil {
		return nil, err
	}
	return json.Marshal(message{
		JSONRPC: jsonrpcVersion,
		Method:  MethodE2EHandshake,
		Route:   &RouteHeader{ChanID: chanID},
		Body:    json.RawMessage(body),
	})
}

func HandshakeFromFrame(f RelayFrame) ([]byte, error) {
	var enc string
	if err := json.Unmarshal(f.Body, &enc); err != nil {
		return nil, fmt.Errorf("api: handshake body: %w", err)
	}
	return base64.StdEncoding.DecodeString(enc)
}
